lib/src/license_detection/confidence.dart

```diff
@@ -0,0 +1,223 @@
+/// Levenshtein word [distance] to the number of tokens in known license.
+@visibleForTesting
+double confidencePercentage(int knownLength, int distance) {
+  if (knownLength == 0) {
+    return 1.0;
+  }
+
+  return 1.0 - (distance / knownLength);
+}
+
+@visibleForTesting
+List<Diff> getDiffs(
+  List<Token> inputTokens,
+  List<Token> knownTokens,
+  MatchRange matchRange,
+) {
+  final dmp = DiffMatchPatch();
+
+  final unknownText = tokensNormalizedValue(inputTokens
+      .skip(matchRange.input.start)
+      .take(matchRange.input.end - matchRange.input.start));
+
+  final knownText = tokensNormalizedValue(
+    knownTokens.take(matchRange.source.end),
+  );
+
+  final diffs = dmp.diffMain(unknownText, knownText);
+
+  dmp.diffCleanupSemantic(diffs);
+
+  return List.unmodifiable(diffs);
+}
+
+/// Returns the range in [diffs] that best resembles the known license text.
+@visibleForTesting
+Range diffRange(String known, List<Diff> diffs) {
+  var foundStart = false;
+  var seen = '';
+  var end = 0;
+  var start = 0;
+
+  for (end = 0; end < diffs.length; end++) {
+    if (seen.length > 1 && seen.trimRight() == known) {
+      break;
+    }
+
+    if (diffs[end].operation == Operation.insert ||
+        diffs[end].operation == Operation.equal) {
+      if (!foundStart) {
+        start = end;
+        foundStart = true;
+      }
+
+      seen += diffs[end].text + ' ';
+    }
+  }
+
+  return Range(start, end);
+}
+
+/// Returns a score indicating to how much extent the [diffs] are acceptable.
+///
+/// If a negative integer is returned it implies that the changes are not
+/// accepatable and hence we discard the known license against which the diffs
+/// were calculated.
+///
+/// If a positive integer is returned it indicates the Levenshtein edit
+/// distance calculated on the number of words altered instead of
+/// the number of characters altered.
+@visibleForTesting
+int scoreDiffs(Iterable<Diff> diffs) {
+  var prevText = '';
+  var prevDelete = '';
+
+  // Make an initial check on the diffs to see if their are any
+  // unacceptable substitutions made.
+  //
+  // As delete diffs are ordered before insert diffs, we store the
+  // previously deleted diff and then compare with insert diff to
+  // check if this substitution is valid or not. If we come across
+  // an invalid substitution return a negative score indicating a
+  // mismatch.
+  for (var diff in diffs) {
+    final text = diff.text;
+
+    switch (diff.operation) {
+      case Operation.insert:
+        var number = text;
+        final index = text.indexOf(' ');
+
+        if (index != -1) {
+          number = number.substring(0, index);
+        }
+
+        // Check if there was a version change and return -1.
+        if (_isVersionNumber(number) && prevText.endsWith('version')) {
+          if (!prevText.endsWith('the standard version') &&
+              !prevText.endsWith('the contributor version')) {
+            return versionChange;
+          }
+        }
```

Comment:
It feels like there are 2 or 3 special cases where that we're trying to handle... And what we're trying to do is to explicitly reject the match, if one of these special cases happens.

Why don't we split these special cases into individual functions. Like:
```dart
void verifyNoVerisonChange(Iterable<Diff> diffs) {
  for (final d in diffs) {
    if (d.operation == Operation.insert) {
      var number = text;
      final index = text.indexOf(' ');

      if (index != -1) {
        number = number.substring(0, index);
      }

      // Check if there was a version change and return -1.
      if (_isVersionNumber(number) && prevText.endsWith('version')) {
        if (!prevText.endsWith('the standard version') &&
            !prevText.endsWith('the contributor version')) {
          throw LicenseMismatchException('License match rejected due to version change!');
        }
      }
    }
  }
}
```

Then we don't have to return negative from `scoreDiffs`, instead we just call `verifyNoVerisonChange` from inside `scoreDiffs` and document that `scoreDiffs` will throw `LicenseMismatchException` if the match is known to be wrong.

This might make the code a bit simpler,  and easier to document. And we can separate `verifyNoVerisonChange` from the checks that looks at `lesser` and `gnu gpl`... This makes it easier to write proper documentation for `verifyNoVerisonChange` so that we know why we're doing it. And we can test `verifyNoVerisonChange` separately.

Reply:
I similarly propose that we make a function `verifyNoLesserGnuGplMismatch(Iterable<Diff> diffs)` which also throws `LicenseMismatchException`..

I realize this means that we'll iterate through the list of `diffs` more than once, but that's okay. Iterating through a list is probably fast -- if it turns out to be slow we can reconsider it later, but right now I'm guess this will be very fast.

Reply:
By throw did you mean similar to `argument error` while we were loading files and abort the process. 

If this is what you meant should we really abort the process because of one `LicenseMismatch`, may we can find a match in the remaining `possible licenses and aborting now may prevent us from finding that match, instead can we do something like

```dart
class Result{
List<LicenseMatch> match;
List<LicenseMismatchException> mismatches
}
```
and we can have some other information here like, the amount of time taken for the whole process, etc...